### PR TITLE
Check if ldProgram can create merge objects

### PR DESCRIPTION
**Pull request by erikd**
The key is "ld command can create merge objects" and has a value of "YES" (the default) or "NO" (if the linker is `lld`).

---

**Template Α: This PR modifies `cabal` behaviour**

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [ ] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.
* [ ] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.

Bonus points for added automated tests!

### CODE REVIEW

File: Cabal/src/Distribution/Simple/GHC/Internal.hs [hunk 1]
```diff
@@ -283,9 +283,24 @@ getGhcInfo verbosity _implInfo ghcProg = do
...
     [(i, ss)]
       | all isSpace ss ->
-          return i
+          return $ detectLLd i
     _ ->
       dieWithException verbosity CantParseGHCOutput
+  where
+    -- The `lld` linker cannot create merge objects so we want to add a new key
+    -- to this map specifying whether the linker can create merge objects.
+    detectLLd :: [(String, String)] -> [(String, String)]
```
Comment (andreabedini) [inline Cabal/src/Distribution/Simple/GHC/Internal.hs:292]: I am not familiar with this part of the code base but it looks like we can detect features of the linker when we "configure" it. E.g. `programProperties` in https://hackage.haskell.org/package/Cabal-3.10.1.0/docs/Distribution-Simple-Program-Types.html#t:ConfiguredProgram

I'll check if this helps us.

**Comment by erikd** [comment]
The implementation has changes so that instead of detecting whether the linker is `lld` we now directly test whether the linker supports the `--relocatable` flag (`lld` does not).

File: Cabal/src/Distribution/Simple/Program/Builtin.hs [hunk 2]
```diff
@@ -347,7 +347,23 @@ greencardProgram :: Program
...
+        -- The `lld` linker cannot create merge (relocatable) objects so we
+        -- want to detect this.
+        ldHelpOutput <-
+          getProgramInvocationOutput
+            verbosity
+            (programInvocation ldProg ["--help"])
+            -- In case the linker does not support '--help'. Eg the LLVM linker,
+            -- `lld` only accepts `-help`.
+            `catchIO` (\_ -> return "")
+        let k = "Supports relocatable output"
```
Comment (angerman) [inline Cabal/src/Distribution/Simple/Program/Builtin.hs:362]:
```suggestion
        -- We use ld -r (--relocatable) to create partially pre-linked objects for GHCi, these allow much faster loading,
        -- as we do not have to do the separate loading and in-memory linking the static linker in GHC does, but can
        -- offload parts of this process to a pre-linking step.
        -- However this requires the linker to support this features. Not all linkers do, and notably as of this writing lld
        -- which is a popular choice for windows linking does not support this feature. However if using binutils ld or
        -- another linker that supports --relocatable, we should still be good to generate pre-linked objects.
        let k = "Supports relocatable output"
```

**Review by angerman — APPROVED** [review]
lgtm, we may want to add some extra documentation as to why we do this.

File: Cabal/src/Distribution/Simple/Program/Builtin.hs (hunk 2 above)
Comment (erikd) [reply Cabal/src/Distribution/Simple/Program/Builtin.hs:372]: More extensive comments added.

**Review by geekosaur — APPROVED** [review]

**Comment by erikd** [comment]
Rebased against `master`.

**Pull request merged**